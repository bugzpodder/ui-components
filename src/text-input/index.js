// @flow
export { TextInput } from "./v1/text-input";
export { TextInputV2 } from "./v2/text-input";
