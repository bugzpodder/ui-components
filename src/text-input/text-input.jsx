// @flow
import React, { type ElementConfig } from "react";
import TextField from "@material-ui/core/TextField";
import { ReadOnlyTextField } from "../readonly-text-field";

type Props = {
  /** If true, displays a read-only input field (`ReadOnlyTextField`). */
  readOnly?: boolean,
  /** If readOnly is true, display this component rather than ReadOnlyTextField. */
  ReadOnlyComponent?: Node<*>,
  /** props for the ReadOnlyComponent */
  readOnlyComponentProps?: Object,
  /** Used for read only input field, see `ReadOnlyTextField`. */
  showEmptyValue?: boolean,
  /** Id of the element. */
  id?: string,
  /** Classes to pass to the element. */
  className?: string,
} & ElementConfig<typeof TextField>;

/** `TextInput` is a wrapper around Material UI's `TextField` which allows `readOnly` as a property. */
export const TextInput = (props: Props) => {
  const {
    error = false,
    helperText = "",
    readOnly,
    ReadOnlyComponent,
    readOnlyComponentProps = {},
    showEmptyValue = false,
    id,
    className,
    ...textFieldProps
  } = props;
  if (readOnly) {
    if (ReadOnlyComponent) {
      return (
        <ReadOnlyComponent
          id={id}
          {...readOnlyComponentProps}
        >
          {props.value}
        </ReadOnlyComponent>
      );
    }
    return (
      <ReadOnlyTextField
        showEmptyValue={showEmptyValue}
        id={id}
        className={className}
      >
        {props.value}
      </ReadOnlyTextField>
    );
  }
  return (
    <TextField
      data-testid="text-field"
      error={error}
      helperText={helperText}
      id={id}
      className={className}
      {...textFieldProps}
    />
  );
};
