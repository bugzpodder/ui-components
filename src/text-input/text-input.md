### Props

All the props of `TextField` are included for `TextInput`.

### Example

```js
const styles = require("../utils/example-styles.module.scss");
class TextInputExample extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      one: "",
      two: "",
      three: "",
    };
    this.handleFirst = this.handleFirst.bind(this);
    this.handleSecond = this.handleSecond.bind(this);
    this.handleThird = this.handleThird.bind(this);
  }

  handleFirst(event) {
    this.setState({ one: event.currentTarget.value });
  }

  handleSecond(event) {
    this.setState({ two: event.currentTarget.value });
  }

  handleThird(event) {
    this.setState({ three: event.currentTarget.value });
  }

  render() {
    return (
      <div className={styles.container}>
        <TextInput
          placeholder="Text Input"
          value={this.state.one}
          onChange={this.handleFirst}
          className={styles.spacing}
        />
        <br />
        <TextInput
          placeholder="Text Input"
          helperText="some helper text"
          value={this.state.two}
          onChange={this.handleSecond}
          className={styles.spacing}
        />
        <br />
        <TextInput
          placeholder="Uh oh"
          helperText="an error"
          error={true}
          value={this.state.three}
          onChange={this.handleThird}
          className={styles.spacing}
        />
        <br />
        <TextInput
          placeholder="Disabled Text Input"
          helperText="you can't type"
          disabled={true}
          className={styles.spacing}
        />
        <br />
        <TextInput placeholder="Text Input" readOnly={true} value="read-only value" className={styles.spacing} />
        <br />
        <TextInput
          readOnly={true}
          ReadOnlyComponent="div"
          readOnlyComponentProps={{ className: "example-classname", style: { color: "red", fontSize: "32px" } }}
          placeholder="Text Field"
          value="read-only value"
          className={styles.spacing}
        />
      </div>
    );
  }
}

<TextInputExample />;
```
