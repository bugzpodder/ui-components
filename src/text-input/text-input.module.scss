.root {
  .disabled {
    background-color: white;
    color: black;
  }
}
