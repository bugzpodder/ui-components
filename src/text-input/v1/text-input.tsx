import React, { ElementType } from "react";
import TextField, { TextFieldProps } from "@material-ui/core/TextField";
import { ReadOnlyTextField } from "../../readonly-text-field";

type Props = {
  /** If true, displays a read-only input field (`ReadOnlyTextField`). */
  readOnly?: boolean;
  /** If readOnly is true, display this component rather than ReadOnlyTextField. */
  ReadOnlyComponent?: ElementType;
  /** props for the ReadOnlyComponent */
  readOnlyComponentProps?: Record<string, any>;
  /** Used for read only input field, see `ReadOnlyTextField`. */
  showEmptyValue?: boolean;
  /** Id of the element. */
  id?: string;
  /** Classes to pass to the element. */
  className?: string;
} & TextFieldProps;

/** `TextInput` is a wrapper around Material UI's `TextField` which allows `readOnly` as a property. */
export const TextInput: React.FC<Props> = props => {
  const {
    error = false,
    helperText = "",
    readOnly,
    ReadOnlyComponent,
    readOnlyComponentProps = {},
    showEmptyValue = false,
    id,
    className,
    ...textFieldProps
  } = props;
  if (readOnly) {
    if (ReadOnlyComponent) {
      return (
        <ReadOnlyComponent id={id} {...readOnlyComponentProps}>
          {props.value}
        </ReadOnlyComponent>
      );
    }
    return (
      <ReadOnlyTextField
        showEmptyValue={showEmptyValue}
        id={id}
        className={className}
      >
        {props.value}
      </ReadOnlyTextField>
    );
  }
  return (
    <TextField
      data-testid="text-field"
      error={error}
      helperText={helperText}
      id={id}
      className={className}
      {...textFieldProps}
    />
  );
};
