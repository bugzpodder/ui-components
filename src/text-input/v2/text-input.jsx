// @flow
import React, { type ElementConfig, useState } from "react";
import TextField from "@material-ui/core/TextField";
import styles from "./text-input.module.scss";

const isEmptyValue = value => value == null || value === "";

const getDisplayValue = (
  value,
  readOnly?: boolean,
  hasReadOnlyDefaultValue: boolean,
  readOnlyDefaultValue?: any
) => {
  const defaultValue = hasReadOnlyDefaultValue ? readOnlyDefaultValue : "-";
  return isEmptyValue(value) && readOnly ? defaultValue : value;
};

type Props = {
  /** If true, displays the field as read-only (with label and helper text). */
  readOnly?: boolean,
  /** Custom readOnly default value. Default to "-". It could be set explicitly to empty value, e.g. "". */
  readOnlyDefaultValue?: any,
  /** ID of the element. */
  id?: string,
  /** Classes to pass to the element. */
  className?: string,
} & ElementConfig<typeof TextField>;

/** TextInputV2 is a wrapper around the filled variant of Material UI's
TextField that accepts "readOnly" as a property.
*/
export const TextInputV2 = (props: Props) => {
  const {
    id,
    className,
    value,
    readOnly,
    readOnlyDefaultValue,
    disabled,
    placeholder,
    "data-testid": dataTestId,
    onFocus,
    onBlur,
    ...otherProps
  } = props;
  const [isInputFocused, setIsInputFocused] = useState(false);
  // We check if readOnlyDefaultValue is explicitly set.
  const hasReadOnlyDefaultValue = readOnlyDefaultValue != null;
  const displayValue = getDisplayValue(value, readOnly,
    hasReadOnlyDefaultValue, readOnlyDefaultValue);

  return (
    <TextField
      data-testid={dataTestId || `${readOnly ? "readonly-" : ""}text-input`}
      id={id}
      className={className}
      margin="normal"
      variant="filled"
      value={displayValue}
      disabled={readOnly || disabled}
      placeholder={readOnly && hasReadOnlyDefaultValue ? undefined : placeholder}
      InputProps={{
        classes: {
          root: styles.root,
          disabled: styles.disabled,
        },
        disableUnderline: readOnly,
        readOnly,
        onFocus: event => {
          onFocus && onFocus(event);
          setIsInputFocused(true);
        },
        onBlur: event => {
          onBlur && onBlur(event);
          setIsInputFocused(false);
        },
      }}
      InputLabelProps={{
        shrink: readOnly || !isEmptyValue(value) || isInputFocused,
      }}
      {...otherProps}
    />
  );
};
