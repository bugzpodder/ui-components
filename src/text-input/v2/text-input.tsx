import React, { useState } from "react";
import TextField, { TextFieldProps } from "@material-ui/core/TextField";
import styles from "./text-input.module.scss";

const isEmptyValue = (value): boolean => value == null || value === "";

const getDisplayValue = (
  value,
  readOnly: boolean,
  hasReadOnlyDefaultValue: boolean,
  readOnlyDefaultValue?: any,
): string => {
  const defaultValue = hasReadOnlyDefaultValue ? readOnlyDefaultValue : "-";
  return isEmptyValue(value) && readOnly ? defaultValue : value;
};

type Props = {
  /** If true, displays the field as read-only (with label and helper text). */
  readOnly?: boolean;
  /** Custom readOnly default value. Default to "-". It could be set explicitly to empty value, e.g. "". */
  readOnlyDefaultValue?: any;
  /** ID of the element. */
  id?: string;
  /** Classes to pass to the element. */
  className?: string;
  /** Data test id */
  "data-testid"?: string;
} & TextFieldProps;

/** TextInputV2 is a wrapper around the filled variant of Material UI's
TextField that accepts "readOnly" as a property.
*/
export const TextInputV2: React.FC<Props> = props => {
  const {
    id,
    className,
    value,
    readOnly,
    readOnlyDefaultValue,
    disabled,
    placeholder,
    "data-testid": dataTestId,
    onFocus,
    onBlur,
    ...otherProps
  } = props;
  const [isInputFocused, setIsInputFocused] = useState(false);
  // We check if readOnlyDefaultValue is explicitly set.
  const hasReadOnlyDefaultValue = readOnlyDefaultValue != null;
  const displayValue = getDisplayValue(
    value,
    readOnly,
    hasReadOnlyDefaultValue,
    readOnlyDefaultValue,
  );

  return (
    // @ts-ignore: disableUnderline not found on type
    <TextField
      data-testid={dataTestId || `${readOnly ? "readonly-" : ""}text-input`}
      id={id}
      className={className}
      margin="normal"
      variant="filled"
      value={displayValue}
      disabled={readOnly || disabled}
      placeholder={
        readOnly && hasReadOnlyDefaultValue ? undefined : placeholder
      }
      InputProps={{
        classes: {
          root: styles.root,
          disabled: styles.disabled,
        },
        disableUnderline: readOnly,
        readOnly,
        onFocus: event => {
          onFocus && onFocus(event);
          setIsInputFocused(true);
        },
        onBlur: event => {
          onBlur && onBlur(event);
          setIsInputFocused(false);
        },
      }}
      InputLabelProps={{
        shrink: readOnly || !isEmptyValue(value) || isInputFocused,
      }}
      {...otherProps}
    />
  );
};
