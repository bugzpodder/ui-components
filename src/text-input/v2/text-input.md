### Props

All the props of `TextInput` are included for `TextInputV2`.

### Example

```js
import styles from "../../test-utils/example-styles.module.scss";
import { ExampleWrapper } from "../../test-utils";
import { useState } from "react";
import { TextInputV2 } from "./text-input";

const TextInputV2Example = () => {
  const [one, setOne] = useState("");
  const [two, setTwo] = useState("");
  const [three, setThree] = useState("");
  const [four, setFour] = useState("");

  const handleFirst = event => {
    setOne(event.currentTarget.value);
  };

  const handleSecond = event => {
    setTwo(event.currentTarget.value);
  };

  const handleThird = event => {
    setThree(event.currentTarget.value);
  };

  const handleFourth = event => {
    setFour(event.currentTarget.value);
  };

  return (
    <div className={styles.container}>
      <TextInputV2
        label="some label"
        placeholder="Filled Text Input"
        value={one}
        onChange={handleFirst}
        className={styles.spacing}
      />
      <TextInputV2
        label="some label"
        placeholder="Filled Text Input"
        helperText="some helper text"
        value={two}
        onChange={handleSecond}
        className={styles.spacing}
      />
      <TextInputV2
        label="some disabled input"
        placeholder="Disabled"
        helperText="some helper text"
        disabled={true}
        className={styles.spacing}
      />
      <TextInputV2
        placeholder="Uh oh"
        helperText="an error"
        error={true}
        value={three}
        onChange={handleThird}
        className={styles.spacing}
      />
      <br />
      <TextInputV2
        label="some label"
        readOnly={true}
        placeholder="Placeholder text, won't show"
        value="filled read-only value"
        data-testid="fourth-text-input"
        className={styles.spacing}
      />
      <TextInputV2
        label="some label"
        readOnly={true}
        placeholder="Placeholder text, won't show"
        helperText="default value"
        value=""
        className={styles.spacing}
      />
      <TextInputV2
        label="some label"
        readOnly={true}
        readOnlyDefaultValue="custom default value"
        placeholder="Placeholder text, won't show"
        value=""
        className={styles.spacing}
      />
      <TextInputV2
        label="some label"
        readOnly={true}
        readOnlyDefaultValue=""
        helperText="read only show empty value"
        placeholder="Placeholder text, won't show"
        value=""
        className={styles.spacing}
      />
    </div>
  );
};

<ExampleWrapper>
  <TextInputV2Example />
</ExampleWrapper>;
```
