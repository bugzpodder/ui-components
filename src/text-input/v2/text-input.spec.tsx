import React from "react";
import { TestWrapper } from "../../test-utils";
import { cleanup, fireEvent, render } from "@testing-library/react";

import "@testing-library/jest-dom/extend-expect";
import { TextInputV2 } from "./text-input";

afterEach(cleanup);

test("render text field wrapper", async () => {
  const testValue = "test value";
  const testPlaceholder = "Test Placeholder";
  const { container, getByPlaceholderText } = render(
    <TestWrapper>
      <TextInputV2
        placeholder={testPlaceholder}
        id="operator"
        onChange={() => {}}
        value={testValue}
      />
    </TestWrapper>,
  );
  expect(getByPlaceholderText(testPlaceholder)).toHaveAttribute(
    "value",
    testValue,
  );
  expect(container).toMatchSnapshot();
});

test("render text field wrapper with label, helperText, and readonly", async () => {
  const testValue = "test value";
  const testLabel = "test label";
  const testHelperText = "test helper text";
  const { container, getByTestId, getByDisplayValue } = render(
    <TestWrapper>
      <TextInputV2
        label={testLabel}
        id="operator"
        readOnly
        onChange={() => {}}
        value={testValue}
        helperText={testHelperText}
      />
    </TestWrapper>,
  );
  // This will error out if test value doesn't show up.
  expect(getByDisplayValue(testValue)).toBeDisabled();
  expect(getByTestId("readonly-text-input")).toHaveTextContent(testLabel);
  expect(getByTestId("readonly-text-input")).toHaveTextContent(testHelperText);
  expect(container).toMatchSnapshot();
});

test("render text field wrapper with readonly as true show default value", async () => {
  const testValue = "";
  const { container, getByDisplayValue } = render(
    <TestWrapper>
      <TextInputV2
        id="operator"
        onChange={() => {}}
        value={testValue}
        readOnly
      />
    </TestWrapper>,
  );
  // This will error out if test value doesn't show up.
  expect(getByDisplayValue("-")).toBeDisabled();
  expect(container).toMatchSnapshot();
});

test("text field wrapper with readonly as true shows custom default value", async () => {
  const testValue = "";
  const readOnlyDefaultValue = "blahhh";
  const { container, getByDisplayValue } = render(
    <TestWrapper>
      <TextInputV2
        id="operator"
        onChange={() => {}}
        value={testValue}
        readOnly
        readOnlyDefaultValue={readOnlyDefaultValue}
      />
    </TestWrapper>,
  );
  // This will error out if test value doesn't show up.
  expect(getByDisplayValue(readOnlyDefaultValue)).toBeDisabled();
  expect(container).toMatchSnapshot();
});

test("text field wrapper with readonly as true shows custom default value", async () => {
  const testValue = "";
  const readOnlyDefaultValue = "";
  const { container, getByDisplayValue } = render(
    <TestWrapper>
      <TextInputV2
        id="operator"
        onChange={() => {}}
        value={testValue}
        readOnly
        readOnlyDefaultValue={readOnlyDefaultValue}
        placeholder="this won't show"
      />
    </TestWrapper>,
  );
  // This will error out if test value doesn't show up.
  expect(getByDisplayValue(testValue)).toBeDisabled();
  expect(() => getByDisplayValue("this won't show")).toThrow();
  expect(container).toMatchSnapshot();
});

test("passed down onBlur and onFocus are called", async () => {
  const testValue = "hi";
  const mockBlur = jest.fn(() => {});
  const mockFocus = jest.fn(() => {});
  const { container, getByDisplayValue } = render(
    <TestWrapper>
      <TextInputV2
        id="operator"
        onChange={() => {}}
        onBlur={mockBlur}
        onFocus={mockFocus}
        value={testValue}
      />
    </TestWrapper>,
  );
  fireEvent.focus(getByDisplayValue(testValue));
  expect(mockFocus.mock.calls.length).toBe(1);
  fireEvent.blur(getByDisplayValue(testValue));
  expect(mockBlur.mock.calls.length).toBe(1);
  expect(container).toMatchSnapshot();
});
