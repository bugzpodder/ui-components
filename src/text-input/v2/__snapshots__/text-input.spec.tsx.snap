// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passed down onBlur and onFocus are called 1`] = `
<div>
  <div
    class=""
    data-testid="text-input"
  >
    <div
      class="root"
    >
      <input
        aria-invalid="false"
        class=""
        id="operator"
        type="text"
        value="hi"
      />
    </div>
  </div>
</div>
`;

exports[`render text field wrapper 1`] = `
<div>
  <div
    class=""
    data-testid="text-input"
  >
    <div
      class="root"
    >
      <input
        aria-invalid="false"
        class=""
        id="operator"
        placeholder="Test Placeholder"
        type="text"
        value="test value"
      />
    </div>
  </div>
</div>
`;

exports[`render text field wrapper with label, helperText, and readonly 1`] = `
<div>
  <div
    class=""
    data-testid="readonly-text-input"
  >
    <label
      class=""
      data-shrink="true"
      for="operator"
      id="operator-label"
    >
      test label
    </label>
    <div
      class="root disabled"
    >
      <input
        aria-describedby="operator-helper-text"
        aria-invalid="false"
        class="disabled"
        disabled=""
        id="operator"
        readonly=""
        type="text"
        value="test value"
      />
    </div>
    <p
      class=""
      id="operator-helper-text"
    >
      test helper text
    </p>
  </div>
</div>
`;

exports[`render text field wrapper with readonly as true show default value 1`] = `
<div>
  <div
    class=""
    data-testid="readonly-text-input"
  >
    <div
      class="root disabled"
    >
      <input
        aria-invalid="false"
        class="disabled"
        disabled=""
        id="operator"
        readonly=""
        type="text"
        value="-"
      />
    </div>
  </div>
</div>
`;

exports[`text field wrapper with readonly as true shows custom default value 1`] = `
<div>
  <div
    class=""
    data-testid="readonly-text-input"
  >
    <div
      class="root disabled"
    >
      <input
        aria-invalid="false"
        class="disabled"
        disabled=""
        id="operator"
        readonly=""
        type="text"
        value="blahhh"
      />
    </div>
  </div>
</div>
`;

exports[`text field wrapper with readonly as true shows custom default value 2`] = `
<div>
  <div
    class=""
    data-testid="readonly-text-input"
  >
    <div
      class="root disabled"
    >
      <input
        aria-invalid="false"
        class="disabled"
        disabled=""
        id="operator"
        readonly=""
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
